File: 0073_Set_Matrix_Zeroes/solution.java
package cn.huang.leetcode;

/*
73. Set Matrix Zeroes
Given a m x n matrix, if an element is 0, set its entire row and column to 0. Do it in-place.

Example 1:

Input:
[
  [1,1,1],
  [1,0,1],
  [1,1,1]
]
Output:
[
  [1,0,1],
  [0,0,0],
  [1,0,1]
]

Example 2:

Input:
[
  [0,1,2,0],
  [3,4,5,2],
  [1,3,1,5]
]
Output:
[
  [0,0,0,0],
  [0,4,5,0],
  [0,3,1,0]
]

Follow up:

    A straight forward solution using O(mn) space is probably a bad idea.
    A simple improvement uses O(m + n) space, but still not the best solution.
    Could you devise a constant space solution?


 */
public class LC_0073_SetMatrixZeroes {
    /*
        Space (m + m)
     */
    public void setZeroes(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        boolean[] zerorow = new boolean[rows];
        boolean[] zerocolumn = new boolean[cols];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == 0) {
                    zerorow[i] = true;
                    zerocolumn[j] = true;
                }
            }
        }

        for (int i = 0; i < rows; i++) {
            if (zerorow[i]) {
                for (int j = 0; j < cols; j++) {
                    matrix[i][j] = 0;
                }
            }
        }

        for (int j = 0; j < cols; j++) {
            if (zerocolumn[j]) {
                for (int i = 0; i < rows; i++) {
                    matrix[i][j] = 0;
                }
            }
        }
    }

    /*
    Can not pass
    [[2147483647],[2],[3]]
     */
    public void setZeroes_MaxInt(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == 0) {
                    for (int x = 0; x < cols; x++) {
                        if (matrix[i][x] != 0) matrix[i][x] = Integer.MAX_VALUE;
                    }
                    for (int y = 0; y < rows; y++) {
                        if (matrix[y][j] != 0) matrix[y][j] = Integer.MAX_VALUE;
                    }
                }
            }
        }

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == Integer.MAX_VALUE) {
                    matrix[i][j] = 0;
                }
            }
        }
    }

    /*
    Step 1:
    First row contains zero = true;
    First column contains zero = false;

    Step 2: use first row and column to mark zero row and column.
    set-matrix-zero-2

    Step 3: set each elements by using marks in first row and column.
    set-matrix-zero-3

    Step 4: Set first column and row by using marks in step 1.
     */
    public void setZeroes_Constant(int[][] matrix) {
        boolean firstRowZero = false;
        boolean firstColumnZero = false;

        //set first row and column zero or not
        for (int i = 0; i < matrix.length; i++) {
            if (matrix[i][0] == 0) {
                firstColumnZero = true;
                break;
            }
        }

        for (int i = 0; i < matrix[0].length; i++) {
            if (matrix[0][i] == 0) {
                firstRowZero = true;
                break;
            }
        }

        //mark zeros on first row and column
        for (int i = 1; i < matrix.length; i++) {
            for (int j = 1; j < matrix[0].length; j++) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }

        //use mark to set elements
        for (int i = 1; i < matrix.length; i++) {
            for (int j = 1; j < matrix[0].length; j++) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }

        //set first column and row
        if (firstColumnZero) {
            for (int i = 0; i < matrix.length; i++)
                matrix[i][0] = 0;
        }

        if (firstRowZero) {
            for (int i = 0; i < matrix[0].length; i++)
                matrix[0][i] = 0;
        }
    }
}

File: 0073_Set_Matrix_Zeroes/solution.cs
/*
73. Set Matrix Zeroes

Given a m x n matrix, if an element is 0, set its entire row and column to 0. Do it in place.

Follow up:
Did you use extra space?
A straight forward solution using O(mn) space is probably a bad idea.
A simple improvement uses O(m + n) space, but still not the best solution.
Could you devise a constant space solution?

*/
using System;

public class Solution073 {
    public void SetZeroes_colrow(int[,] matrix) {
        var row = matrix.GetLength(0);
        var column = matrix.GetLength(1);

        var zerorow = new bool[row];
        var zerocolumn = new bool[column];

        for (int i = 0; i < row; i++)
        {
            for (int j = 0; j < column; j++)
            {
                if (matrix[i, j] == 0)
                {
                    zerorow[i] = true;
                    zerocolumn[j] = true;
                }
            }
        }

        for (int i = 0; i < row; i++)
        {
            if (zerorow[i])
            {
                for (int j = 0; j < column; j++)
                {
                    matrix[i, j] = 0;
                }
            }
        }

        for (int j = 0; j < column; j++)
        {
            if (zerocolumn[j])
            {
                for (int i = 0; i < row; i++)
                {
                    matrix[i, j] = 0;
                }
            }
        }
    }

    public void SetMatrixZeroes_MaxInt(int[][] matrix)
    {
        var m = matrix.Length;
        var n = matrix[0].Length;

        for (var i = 0; i < m; i++)
        {
            for (var j = 0; j < n; j++)
            {
                if (matrix[i][j] == 0)
                {
                    for (var x = 0; x < n; x++)
                    {
                        if (matrix[i][x] != 0) matrix[i][x] = Int32.MaxValue;
                    }
                    for (var y = 0; y < m; y++)
                    {
                        if (matrix[y][j] != 0) matrix[y][j] = Int32.MaxValue;
                    }
                }
            }
        }

        for (var i = 0; i < m; i++)
        {
            for (var j = 0; j < n; j++)
            {
                if (matrix[i][j] == Int32.MaxValue)
                {
                    matrix[i][j] = 0;
                }
            }
        }
    }
}